Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,634 @@
+}
+
+void TextAreaBox::ElidedLabel::set_alignment(Qt::Alignment alignment) {
+  m_alignment = alignment;
+  update();
+}
+
+void TextAreaBox::ElidedLabel::set_line_height(int line_height) {
+  m_line_height = line_height;
+  update();
+}
+
+QSize TextAreaBox::ElidedLabel::sizeHint() const  {
+  return {width(), height()};
+}
+
+void TextAreaBox::ElidedLabel::paintEvent(QPaintEvent *event) {
+  auto painter = QPainter(this);
+  painter.setPen(m_text_color);
+  painter.setFont(font());
+  auto fontMetrics = painter.fontMetrics();
+  auto y = m_line_height - fontMetrics.height();
+  auto textLayout = QTextLayout(m_text, painter.font());
+  auto opt = textLayout.textOption();
+  opt.setAlignment(m_alignment);
+  textLayout.setTextOption(opt);
+  textLayout.beginLayout();
+  auto line = textLayout.createLine();
+  while(line.isValid()) {
+    line.setLineWidth(width());
+    auto next_line_y = y + m_line_height;
+    if (height() >= next_line_y + m_line_height) {
+      line.draw(&painter, QPoint(0, y));
+      y = next_line_y;
+    } else {
+      auto last_line = m_text.mid(line.textStart());
+      auto elided_last_line = fontMetrics.elidedText(last_line,
+        Qt::ElideRight, width());
+      painter.drawText(QPoint(0, y + fontMetrics.ascent()),
+        elided_last_line);
+      line = textLayout.createLine();
+      break;
+    }
+    line = textLayout.createLine();
+  }
+  textLayout.endLayout();
+}
+
+void TextAreaBox::StyleProperties::clear() {
+  m_styles.clear();
+  m_border_sizes = {};
+  m_padding = {};
+  m_alignment = none;
+  m_font = none;
+  m_size = none;
+  m_color = none;
+}
+
+TextAreaBox::TextAreaBox(QWidget* parent)
+  : TextAreaBox(std::make_shared<LocalTextModel>(), parent) {}
+
+TextAreaBox::TextAreaBox(QString current, QWidget* parent)
+  : TextAreaBox(std::make_shared<LocalTextModel>(std::move(current)),
+      parent) {}
+
+TextAreaBox::TextAreaBox(std::shared_ptr<TextModel> model, QWidget* parent)
+    : QWidget(parent),
+      m_placeholder_styles{[=] { commit_placeholder_style(); }},
+      m_text_edit_styles{[=] { commit_style(); }},
+      m_model(std::move(model)),
+      m_submission(m_model->get_current()) {
+  m_text_edit = new ContentSizedTextEdit(m_model->get_current(), this);
+  m_text_edit->installEventFilter(this);
+  m_stacked_widget = new QStackedWidget(this);
+  m_stacked_widget->addWidget(m_text_edit);
+  setFocusProxy(m_text_edit);
+  connect(m_text_edit->document(), &QTextDocument::contentsChanged, this,
+    &TextAreaBox::on_text_changed);
+  m_placeholder = new ElidedLabel(this);
+  m_placeholder->setFixedSize(0, 0);
+  m_stacked_widget->addWidget(m_placeholder);
+  m_placeholder->setCursor(m_text_edit->cursor());
+  m_placeholder->setAttribute(Qt::WA_TransparentForMouseEvents);
+  m_scroll_box = new ScrollBox(m_stacked_widget, this);
+  m_scroll_box->set(ScrollBox::DisplayPolicy::NEVER,
+    ScrollBox::DisplayPolicy::ON_OVERFLOW);
+  m_scroll_box->get_vertical_scroll_bar().adjustSize();
```

Comment:
What is the purpose of this ```adjustSize``` call?

Reply:
In update_text_edit_width(), it's possible that the scroll bar will be shown after the text edit's width is updated, but it's not currently visible. This adjustSize call makes sure that when width() is called that it returns the size the scroll bar would be if it was shown.

On reviewing it, I can move this call to right before the width is queried.

Reply:
If you want what the width would or "should" be, use the component's size hint. ```adjustSize``` is a very expensive call that produces size effects, ```sizeHint``` produces no side effects and is generally an ```O(1)``` operation.